onnxruntime/core/graph/graph_utils.cc

```diff
@@ -445,6 +505,65 @@ size_t RemoveNodeOutputEdges(Graph& graph, Node& node) {
   return output_edges.size();
 }
 
-}  // namespace graph_utils
+void SubstituteNodeOutput(Graph& graph, Node& node, Node& replacement, int replacement_output_idx) {
+  std::vector<GraphEdge> output_edges = GetNodeOutputEdges(node);
 
+  if (!output_edges.empty()) {
+    // all edges should have the same source index so double-check we don't do an invalid replacement
+    const auto output_idx = output_edges[0].src_arg_index;
+
+    const auto& replacement_name = replacement.MutableOutputDefs()[replacement_output_idx]->Name();
+
+    // Remove the output edges of the node first
+    RemoveGraphEdges(graph, output_edges);
+
+    // Create connections between the replacement node and the outgoing nodes
+    for (const auto& output_edge : output_edges) {
+      ORT_ENFORCE(output_edge.src_arg_index == output_idx, "Invalid usage. Node produces multiple different outputs. ",
+                  node.Name());
+
+      // Take care of subgraph inputs.
+      if (OutputEdgeProvidesImplicitInput(graph, output_edge)) {
+        Node& mutable_output_edge_node = *graph.GetNode(output_edge.dst_node);
+        UpdateImplicitInputNameInSubgraph(mutable_output_edge_node, output_edge.arg_name, replacement_name);
+      }
+
+      // Add new edge connecting the input with the output nodes directly.
+      // This also updates the destination node's input node args
+      graph.AddEdge(replacement.Index(), output_edge.dst_node, replacement_output_idx, output_edge.dst_arg_index);
+    }
+  }
+}
+
+void ReplaceNodeInputWithNodeArg(Node& target, int target_input_idx, NodeArg& replacement) {
+  size_t dst_arg_idx = static_cast<size_t>(target_input_idx);
+  auto num_explicit_inputs = target.InputDefs().size();
+
+  if (dst_arg_idx < num_explicit_inputs) {
+    target.MutableInputDefs()[target_input_idx] = &replacement;
+  } else if (dst_arg_idx < num_explicit_inputs + target.ImplicitInputDefs().size()) {
+    // If we need to update an implicit input.
+    target.MutableImplicitInputDefs()[dst_arg_idx - num_explicit_inputs] = &replacement;
+  } else {
+    // logic error in our code
+    ORT_THROW("Invalid input index for node ", target.Name(), ". Index:", target_input_idx,
+              " ExplicitInputs:", num_explicit_inputs,
+              " ImplicitInputs:", target.ImplicitInputDefs().size());
+  }
+}
+
+void FinalizeNodeFusion(Graph& graph, Node& first_node, Node& second_node, Node* replacement_node) {
+  graph_utils::RemoveNodeOutputEdges(graph, first_node);
+  graph_utils::MoveOutput(graph, second_node, replacement_node ? *replacement_node : first_node);
+
```

Comment:
Given we are removing nodes here, isn't any of the checks we do at the CanRemoveNodeAndMergeEdges needed here as well?

Reply:
The (existing) assumption is that the fusion implementation checked that previously. Can add new checks if that's necessary.

Reply:
We can mention in the comment that we assume CanRemoveNodeAndMergeEdges is called before to perform all necessary checks